ipns.go

```diff
@@ -25,6 +33,32 @@ func (s *Shell) Publish(node string, value string) error {
 	return nil
 }
 
+// PublishWithDetails is used for fine grained control over record publishing
+func (s *Shell) PublishWithDetails(contentHash, key string, lifetime, ttl time.Duration, resolve bool) (*PublishResponse, error) {
+	var pubResp PublishResponse
+	args := []string{contentHash}
+	req := s.newRequest(context.Background(), "name/publish", args...)
+	if key == "" {
+		key = "self"
+	}
+	req.Opts["key"] = key
+	req.Opts["lifetime"] = lifetime.String()
```

Comment:
I'd wrap this and `ttl` with an if and only set this when duration is > 0

go-ipfs cares whether or not a option is present - https://github.com/ipfs/go-ipfs/blob/master/core/commands/publish.go#L117